state/watcher.go

```diff
@@ -1096,150 +1095,291 @@ func (w *relationUnitsWatcher) loop() (err error) {
-	commonWatcher
-	members         bson.D
-	filter          func(key interface{}) bool
-	knownLifeStatus map[string]lifeStatus
-	out             chan []string
-}
-
-// WatchStatus returns a watcher that notifies of changes to the life
+// WatchLifeStatus returns a watcher that notifies of changes to the life
 // or status of the relation.
-func (r *Relation) WatchStatus() StringsWatcher {
+func (r *Relation) WatchLifeStatus() StringsWatcher {
 	filter := func(id interface{}) bool {
 		k, err := r.st.strictLocalID(id.(string))
 		if err != nil {
 			return false
 		}
 		return k == r.Tag().Id()
 	}
-	return newRelationStatusWatcher(r.st, nil, filter)
+	members := bson.D{{"id", r.Id()}}
+	lw := newLifecycleWatcher(r.st, relationsC, members, filter, nil)
+
+	return newRelationLifeStatusWatcher(r.st, lw)
+}
+
+// relationLifeStatusWatcher sends notifications of changes to the life or
+// status of specific relations.
+type relationLifeStatusWatcher struct {
+	commonWatcher
+	lifeWatcher StringsWatcher
+	out         chan []string
+
+	// statusWatchers holds status watchers keyed on relation id.
+	statusWatchers map[int]StringsWatcher
+
+	// relationKeys maps global relation scope id to relation key.
+	relationKeys map[string]string
+
+	// statusOut is the channel used to receive status changes.
+	statusOut chan []string
+}
+
+// newRelationLifeStatusWatcher creates a watcher that sends changes when the specified
+// lifeWatcher fires, or when relations which are alive according the the life watcher
+// change status.
+func newRelationLifeStatusWatcher(backend modelBackend, lifeWatcher StringsWatcher) *relationLifeStatusWatcher {
+	w := &relationLifeStatusWatcher{
+		commonWatcher:  newCommonWatcher(backend),
+		lifeWatcher:    lifeWatcher,
+		statusWatchers: make(map[int]StringsWatcher),
+		relationKeys:   make(map[string]string),
+		statusOut:      make(chan []string),
+		out:            make(chan []string),
+	}
+	go func() {
+		defer w.finish()
+		w.tomb.Kill(w.loop())
+	}()
+	return w
+}
+
+func (w *relationLifeStatusWatcher) finish() {
+	watcher.Stop(w.lifeWatcher, &w.tomb)
+	for _, sw := range w.statusWatchers {
+		watcher.Stop(sw, &w.tomb)
+	}
+	close(w.out)
+	w.tomb.Done()
+}
+
+func (w *relationLifeStatusWatcher) Changes() <-chan []string {
+	return w.out
+}
+
+type relationLifeDoc struct {
+	DocId string `bson:"_id"`
+	Life  Life   `bson:"life"`
+	Id    int    `bson:"id"`
+}
+
+var relationLifeFields = bson.D{{"_id", 1}, {"life", 1}, {"id", 1}}
+
+// lifeChanged starts or stops status watchers for relations with the
+// specified keys, according to whether the relation is alive or not.
+func (w *relationLifeStatusWatcher) lifeChanged(keys []string) error {
+	relColl, closer := w.db.GetCollection(relationsC)
+	defer closer()
+
+	iter := relColl.Find(bson.D{{"_id", bson.D{{"$in", keys}}}}).Select(relationLifeFields).Iter()
+	var doc relationLifeDoc
+	for iter.Next(&doc) {
+		rid := relationGlobalScope(doc.Id)
+		switch doc.Life {
+		case Alive:
```

Comment:
I was thinking not since once the relation goes not alive, the status isn't used for anything in practice. I guess it won't hurt to keep the watcher until it goes dead.